llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1835,96 +1839,79 @@ static Instruction *foldSelectICmpEq(SelectInst &SI, ICmpInst *ICI,
-          matchFalseVal(AndOps, AndOps, NotInner))
-        return IC.replaceInstUsesWith(SI, FalseVal);
-    }
-  }
+    // (X | Y) == 0 ? X Op0 Y : X Op1 Y -> X Op1 Y
+    // For any `Op0` and `Op1` that are zero when `X` and `Y` are zero.
+	// https://alive2.llvm.org/ce/z/azHzBW
+    if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y))) &&
+        (match(TrueVal, m_c_BinOp(m_Specific(X), m_Specific(Y))) ||
+		 // In true arm we can also accept just `0`.
+         match(TrueVal, m_Zero())) &&
+        match(FalseVal, m_c_BinOp(m_Specific(X), m_Specific(Y)))) {
+      auto IsOpcZeroWithZeros = [](Value *V) {
+        auto *I = dyn_cast<Instruction>(V);
+        if (!I)
+          return false;
+        switch (I->getOpcode()) {
+        case Instruction::And:
+        case Instruction::Or:
+        case Instruction::Xor:
+        case Instruction::Mul:
+        case Instruction::Add:
+        case Instruction::Sub:
+        case Instruction::Shl:
+        case Instruction::AShr:
+        case Instruction::LShr:
+          return true;
+        default:
+          return false;
+        }
+      };
 
-  // (X^Y)==C ? X&Y : X|Y -> (X|Y)^C : X|Y  or (X|Y)&~C : X|Y
-  // (X^Y)==C ? X|Y : X&Y -> (X&Y)|C : X&Y  or (X&Y)^ C : X&Y
-  if (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y)))) {
-    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))))) {
-      // (X^Y)==C ? X&Y : (X|Y)^C -> (X|Y)^C
-      // (X^Y)==C ? X&Y : (X|Y)&~C -> (X|Y)&~C
-      if (matchFalseVal(XorOps, OrOps, None) ||
-          matchFalseVal(AndOps, OrOps, NotRHS))
-        return IC.replaceInstUsesWith(SI, FalseVal);
-    } else if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
-      // (X^Y)==C ? (X|Y) : (X&Y)|C -> (X&Y)|C
-      // (X^Y)==C ? (X|Y) : (X&Y)^C -> (X&Y)^C
-      if (matchFalseVal(OrOps, AndOps, None) ||
-          matchFalseVal(XorOps, AndOps, None))
+      if ((match(TrueVal, m_Zero()) || IsOpcZeroWithZeros(TrueVal)) &&
+          IsOpcZeroWithZeros(FalseVal))
         return IC.replaceInstUsesWith(SI, FalseVal);
     }
   }
+  // (X == Y) ? X | Y : X & Y
+  // (X == Y) ? X & Y : X | Y
+  // If `X == Y` then `X == Y == X | Y == X & Y`.
+  // NB: `X == Y` is canonicalization of `(X ^ Y) == 0`.
+  // https://alive2.llvm.org/ce/z/SJskbz
+  X = CmpLHS;
+  Y = CmpRHS;
+  auto MatchOrAnd = m_CombineOr(m_c_Or(m_Specific(X), m_Specific(Y)),
+                                m_c_And(m_Specific(X), m_Specific(Y)));
+  if (match(FalseVal, MatchOrAnd) &&
+      // In the true arm we can also just match `X` or `Y`.
+      (match(TrueVal, MatchOrAnd) || match(TrueVal, m_Specific(X)) ||
+       match(TrueVal, m_Specific(Y)))) {
+    // Can't preserve `or disjoint` here so rebuild.
+    auto *BO = dyn_cast<BinaryOperator>(FalseVal);
+    if (!BO)
+      return nullptr;
 
+    return IC.replaceInstUsesWith(
+        SI, IC.Builder.CreateBinOp(BO->getOpcode(), BO->getOperand(0),
+                                   BO->getOperand(1)));
+  }
```

Comment:
Can we handle this case by changing https://github.com/llvm/llvm-project/blob/3321c2d72ab7757dbdd38bdd99a76d89293dac8a/llvm/lib/Analysis/InstructionSimplify.cpp#L4596-L4612 to compare both simplified values instead? Currently we simplify one and compare against the original other. If we compared both simplified values, I think we should be able to handle this pattern without any specialized code.